README.md
# slick-lightbox

[![CDNJS](https://img.shields.io/cdnjs/v/slick-lightbox.svg)](https://cdnjs.com/libraries/slick-lightbox)

A lightbox wrapper for Ken's amazing slick carousel. Credits go to [kenwheeler](https://github.com/kenwheeler) for doing the [hard work](https://github.com/kenwheeler/slick).

## Setup

### Bower

```
bower install -S slick-lightbox
```

### Ruby on Rails

Use the [slick-lightbox-rails](https://github.com/jwir3/slick-lightbox-rails) gem. Add to Gemfile.

```
gem 'slick-lightbox-rails', '>= 0.0.1'
```

Run bundler.

```
bundle install
```

## Usage

Make sure you include CSS for both slick and slick-lightbox. Example HTML:

```html
<!DOCTYPE html>
<html>
<head>
  <title>slick lightbox</title>
  <meta name="viewport" content="width=device-width">
  <link rel="stylesheet" type="text/css" href="bower_components/slick-carousel/slick/slick.css">
  <link rel="stylesheet" type="text/css" href="bower_components/slick-carousel/slick/slick-theme.css">
  <link rel="stylesheet" type="text/css" href="styles/slick-lightbox.css">
</head>
<body>
  <div>
    <ul>
      <li><a href="http://placekitten.com/1600/1300" data-caption="Lorem ipsum 1600x1300">http://placekitten.com/1600/1300</a></li>
      <li><a href="http://placekitten.com/1630/1280" data-caption="Lorem ipsum 1630x1280">http://placekitten.com/1630/1280</a></li>
      <li><a href="http://placekitten.com/1500/1250" data-caption="Lorem ipsum 1500x1250">http://placekitten.com/1500/1250</a></li>
    </ul>
  </div>
  <script type="text/javascript" src="bower_components/jquery/dist/jquery.js"></script>
  <script type="text/javascript" src="bower_components/slick-carousel/slick/slick.min.js"></script>
  <script type="text/javascript" src="scripts/slick-lightbox.js"></script>
  <script type="text/javascript">
    $(document).ready(function(){
      $('ul').slickLightbox({
        itemSelector: '> li > a'
      });
    });
  </script>
</body>
</html>
```

## Options

Option               | Type     |  Default         | Description
------               | ----     |  -------         | -----------
background           | string   | `rgba(0,0,0,.8)` | Background for the lightbox. Is used directly as a CSS `background` statement so that color and/or an image can be used.
closeOnEscape        | boolean  | `true`           | Should the lightbox close upon pressing ESC?
closeOnBackdropClick | boolean  | `true`           | Should the lightbox close upon clicking on backdrop?
destroyTimeout       | number   | `500`            | How many ms should we wait before actually removing the lightbox from the DOM? The default is used so that a 0.5s opacity transition can take place.
itemSelector         | string   | `a`              | jQuery-like selector for the carousel-to-create items. The items should be anchors with `href` attribute aiming at the desired lightbox image. If that's not the case, you should specify the `src` option.
navigateByKeyboard   | boolean  | `true`           | Can you use keyboard arrows to navigate the carousel?
slick                | object   | `{}`             | Slick options to pass to the carousel.
caption              | various  | `false`          | Should the images be captioned? If `false`, no caption is created. You can pass your own **function** accepting the `element` and `info` attributes (e.g. `function(element, info) { return $(element).doSomething(); }`, with `info` being an `{ index: itemIndex, length: itemCount }` object) or a **string** with the name of the `data-attribute` to be fetched (`my-caption` will get the value of `data-my-caption` attribute).
src                  | various  | `false`          | How to get the image urls? If `false`, no src is taken as the `href` attribute. You can pass your own **function** accepting the element as an attribute (e.g. `function(element) { return $(element).doSomething(); }`) or a **string** with the name of the attribute to be fetched (`src` will get the value of `element.src` attribute).
captionPosition      | various  | `'dynamic'`      | Where should the caption be placed? When set to 'dynamic', the caption is displayed right under the image. When 'bottom', the caption is on the bottom of the screen independent to the image size.
images               | various  | `false`          | When passed an array, slick lightbox doesn't scan the elements for image URLs and uses array's values instead.
useHistoryApi        | boolean  | `false`          | When `true`, opening the lightbox does a `history.pushState`. The plugin then listens for the `popstate` event to close the lightbox.
layouts              | object   |                  | See [Layouts](#layouts)
shouldOpen           | function | `null`           | Given a function, checks the return value before opening. If the function returns a falsy value, the lightbox is not opened upon clicking on the items. Accepts three arguments - the slick lightbox instance as the first, the clicked element as the second and the click event as the last.
imageMaxHeight       | number   | `0.9`            | Maximum height of the lightbox images. Relative to the window height.
lazy                 | boolean  | `false`          | Should the images be lazyloaded? 
lazyPlaceholder      | string   | `null`           | Path to lazyloading placeholder. Defaults to a blank gif.

## Events

Events are triggered on the source element. Example:

```javascript
$(document).ready(function(){
  $('ul').slickLightbox().on({
    'show.slickLightbox': function(){ console.log('A `show.slickLightbox` event triggered.'); },1
    'shown.slickLightbox': function(){ console.log('A `shown.slickLightbox` event triggered.'); },
    'hide.slickLightbox': function(){ console.log('A `hide.slickLightbox` event triggered.'); },
    'hidden.slickLightbox': function(){ console.log('A `hidden.slickLightbox` event triggered.'); }
  });
});
```

Event                | Description
------               | ----
show.slickLightbox   | Triggered when the lightbox is opened.
shown.slickLightbox  | Triggered when the lightbox is opened, after the transitions took place.
hide.slickLightbox   | Triggered when the lightbox is closed.
hidden.slickLightbox | Triggered when the lightbox is closed, after the transitions took place.


## Layouts

The layout is customizable using the `layouts` option object.

- `layouts.closeButton`, which defaults to `<button type="button" class="slick-lightbox-close"></button>`

## Lightbox instance

You can access the lightbox instance through the bound element. It's assigned to `.slickLightbox`. Example:

```javascript
$('#my-element').slickLightbox()
console.log('My slickLightbox instance:', $('#my-element')[0].slickLightbox)
// once a lightbox is opened
console.log('My slick instance:', $('#my-element')[0].slickLightbox.slick)
```

## Contributing

Edit the coffee source please. Use gulp to build.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/). Changelog muster from [olivierlacan/keep-a-changelog](https://github.com/olivierlacan/keep-a-changelog).

## [0.2.12] - 2017-10-10
### Fixed
- Handle slick clones on click. Thanks @IrvingArmenta and @othbert

## [0.2.11] - 2017-10-10
### Fixed
- Fixed images option. Thanks @Vesna031

## [0.2.10] - 2017-06-04
### Added
- Fixed captions. Thanks @AntoineBoulanger

## [0.2.9] - 2017-05-07
### Added
- Added lazyload support. Thanks @avbobylev

## [0.2.8] - 2017-04-03
### Added
- Added the `info` argument to the `getElementCaption` function.
- Blur clicked element so that pressing Enter doesn't cause issues.

## [0.2.7] - 2016-01-02
### Added
- Added the `event` argument to the `shouldOpen` callback.

## [0.2.6] - 2016-12-16
### Added
- Added `.slickLightbox` to the bound element.

## [0.2.5] - 2016-12-09
### Added
- Added the `imageMaxHeight` option.

## [0.2.4] - 2016-10-14
### Removed
- Removed npm-shrinkwrap as there are only dev dependencies.

## [0.2.3] - 2016-10-06
### Added
- Added the `shouldOpen` option.

## [0.2.2] - 2016-09-26
### Added
- Use the same slide index numbers as an 'Existing Slick', fixes #25

## [0.2.1] - 2016-07-29
### Fixed
- fix multiple matching groups using `.each`, fixes #21

## [0.2.0] - 2016-07-29
### Added
- this CHANGELOG.md
- the `layouts` option object for specifying the HTML code used for parts of the lightbox, thanks @CHEWX for the idea
- close button layout `layouts.closeButton`


src/scripts/slick-lightbox.js
'use strict';

/*
 * SlickLightbox documentation #

Documentation generated by [CoffeeDoc](http://github.com/omarkhan/coffeedoc)
 */
(function($) {
  var SlickLightbox, defaults;
  SlickLightbox = (function() {

    /*
    		The one and only class used.
     */
    function SlickLightbox(element, options) {
      var that;
      this.options = options;

      /* Binds the plugin. */
      this.element = $(element);
      this.didInit = false;
      that = this;
      this.element.on('click.slickLightbox', this.options.itemSelector, function(e) {
        e.preventDefault();
        return that.init(that.element.find(that.options.itemSelector).index($(this)));
      });
    }

    SlickLightbox.prototype.init = function(index) {

      /* Creates the lightbox, opens it, binds events and calls `slick`. Accepts `index` of the element, that triggered it (so that we know, on which slide to start slick). */
      this.didInit = true;
      this.detectIE();
      this.createModal(index);
      this.bindEvents();
      this.initSlick();
      return this.open();
    };

    SlickLightbox.prototype.createModalItems = function(index) {

      /* Creates individual slides to be used with slick. If `options.images` array is specified, it uses it's contents, otherwise loops through elements' `options.itemSelector`. */
      var a, createItem, links;
      if (this.options.images) {
        links = $.map(this.options.images, function(img) {
          return "<div class=\"slick-lightbox-slick-item\"><div class=\"slick-lightbox-slick-item-inner\"><img class=\"slick-lightbox-slick-img\" src=\"" + img + "\" /></div></div>";
        });
      } else {
        createItem = (function(_this) {
          return function(el) {
            var caption, src;
            caption = _this.getElementCaption(el);
            src = _this.getElementSrc(el);
            return "<div class=\"slick-lightbox-slick-item\"><div class=\"slick-lightbox-slick-item-inner\"><img class=\"slick-lightbox-slick-img\" src=\"" + src + "\" />" + caption + "</div></div>";
          };
        })(this);
        a = this.element.find(this.options.itemSelector);
        if (index === 0 || index === -1) {
          links = $.map(a, createItem);
        } else {
          links = $.map(a.slice(index), createItem);
          $.each(a.slice(0, index), function(i, el) {
            return links.push(createItem(el));
          });
        }
      }
      return links;
    };

    SlickLightbox.prototype.createModal = function(index) {

      /* Creates a `slick`-friendly modal. Rearranges the items so that the `index`-th item is placed first. */
      var html, links;
      links = this.createModalItems(index);
      html = "<div class=\"slick-lightbox slick-hide-init" + (this.isIE ? ' slick-lightbox-ie' : '') + "\" style=\"background: " + this.options.background + ";\">\n	<div class=\"slick-lightbox-inner\">\n		<div class=\"slick-lightbox-slick slick-caption-" + this.options.captionPosition + "\">" + (links.join('')) + "</div>\n		<button type=\"button\" class=\"slick-lightbox-close\"></button>\n	<div>\n<div>";
      this.modalElement = $(html);
      return $('body').append(this.modalElement);
    };

    SlickLightbox.prototype.initSlick = function(index) {

      /* Runs slick by default, using `options.slick` if provided. If `options.slick` is a function, it gets fired instead of us initializing slick. */
      if (this.options.slick != null) {
        if (typeof this.options.slick === 'function') {
          this.options.slick(this.modalElement);
        } else {
          this.slick = this.modalElement.find('.slick-lightbox-slick').slick(this.options.slick);
        }
      } else {
        this.slick = this.modalElement.find('.slick-lightbox-slick').slick();
      }
      return this.modalElement.trigger('init.slickLightbox');
    };

    SlickLightbox.prototype.open = function() {

      /* Opens the lightbox. */
      if (typeof history !== "undefined" && history !== null) {
        if (typeof history.pushState === "function") {
          history.pushState(null, null, '');
        }
      }
      this.element.trigger('show.slickLightbox');
      setTimeout(((function(_this) {
        return function() {
          return _this.element.trigger('shown.slickLightbox');
        };
      })(this)), this.getTransitionDuration());
      return this.modalElement.removeClass('slick-hide-init');
    };

    SlickLightbox.prototype.close = function() {

      /* Closes the lightbox and destroys it, maintaining the original element bindings. */
      this.element.trigger('hide.slickLightbox');
      setTimeout(((function(_this) {
        return function() {
          return _this.element.trigger('hidden.slickLightbox');
        };
      })(this)), this.getTransitionDuration());
      this.modalElement.addClass('slick-hide');
      return this.destroy();
    };

    SlickLightbox.prototype.bindEvents = function() {

      /* Binds global events. */
      var resizeSlides;
      resizeSlides = (function(_this) {
        return function() {
          var h;
          h = _this.modalElement.find('.slick-lightbox-inner').height();
          _this.modalElement.find('.slick-lightbox-slick-item').height(h);
          return _this.modalElement.find('.slick-lightbox-slick-img').css('max-height', Math.round(0.9 * h));
        };
      })(this);
      $(window).on('orientationchange.slickLightbox resize.slickLightbox', resizeSlides);
      $(window).on('popstate.slickLightbox', (function(_this) {
        return function() {
          return _this.close();
        };
      })(this));
      this.modalElement.on('init.slickLightbox', resizeSlides);
      this.modalElement.on('destroy.slickLightbox', (function(_this) {
        return function() {
          return _this.destroy();
        };
      })(this));
      this.element.on('destroy.slickLightbox', (function(_this) {
        return function() {
          return _this.destroy(true);
        };
      })(this));
      this.modalElement.on('click.slickLightbox touchstart.slickLightbox', '.slick-lightbox-close', (function(_this) {
        return function(e) {
          e.preventDefault();
          return _this.close();
        };
      })(this));
      if (this.options.closeOnEscape || this.options.navigateByKeyboard) {
        $(document).on('keydown.slickLightbox', (function(_this) {
          return function(e) {
            var code;
            code = e.keyCode ? e.keyCode : e.which;
            if (_this.options.navigateByKeyboard) {
              if (code === 37) {
                _this.slideSlick('left');
              } else if (code === 39) {
                _this.slideSlick('right');
              }
            }
            if (_this.options.closeOnEscape) {
              if (code === 27) {
                return _this.close();
              }
            }
          };
        })(this));
      }
      if (this.options.closeOnBackdropClick) {
        this.modalElement.on('click.slickLightbox touchstart.slickLightbox', '.slick-lightbox-slick-img', (function(_this) {
          return function(e) {
            return e.stopPropagation();
          };
        })(this));
        return this.modalElement.on('click.slickLightbox touchstart.slickLightbox', '.slick-lightbox-slick-item', (function(_this) {
          return function(e) {
            e.preventDefault();
            return _this.close();
          };
        })(this));
      }
    };

    SlickLightbox.prototype.slideSlick = function(direction) {

      /* Moves the slick prev or next. */
      if (direction === 'left') {
        return this.slick.slick('slickPrev');
      } else {
        return this.slick.slick('slickNext');
      }
    };

    SlickLightbox.prototype.detectIE = function() {

      /* Detects usage of IE8 and lower. */
      var ieversion;
      this.isIE = false;
      if (/MSIE (\d+\.\d+);/.test(navigator.userAgent)) {
        ieversion = new Number(RegExp.$1);
        if (ieversion < 9) {
          return this.isIE = true;
        }
      }
    };

    SlickLightbox.prototype.getElementCaption = function(el) {

      /* Returns caption for each slide based on the type of `options.caption`. */
      var c;
      if (!this.options.caption) {
        return '';
      }
      c = (function() {
        switch (typeof this.options.caption) {
          case 'function':
            return this.options.caption(el);
          case 'string':
            return $(el).data(this.options.caption);
        }
      }).call(this);
      return "<span class=\"slick-lightbox-slick-caption\">" + c + "</span>";
    };

    SlickLightbox.prototype.getElementSrc = function(el) {

      /* Returns src for each slide image based on the type of `options.src`. */
      switch (typeof this.options.src) {
        case 'function':
          return this.options.src(el);
        case 'string':
          return $(el).attr(this.options.src);
        default:
          return el.href;
      }
    };

    SlickLightbox.prototype.unbindEvents = function() {

      /* Unbinds global events. */
      $(window).off('.slickLightbox');
      $(document).off('.slickLightbox');
      return this.modalElement.off('.slickLightbox');
    };

    SlickLightbox.prototype.destroy = function(unbindAnchors) {
      if (unbindAnchors == null) {
        unbindAnchors = false;
      }

      /* Destroys the lightbox and unbinds global events. If `true` is passed as an argument, unbinds the original element as well. */
      if (this.didInit) {
        this.unbindEvents();
        setTimeout(((function(_this) {
          return function() {
            return _this.modalElement.remove();
          };
        })(this)), this.options.destroyTimeout);
      }
      if (unbindAnchors) {
        this.element.off('.slickLightbox');
        return this.element.off('.slickLightbox', this.options.itemSelector);
      }
    };

    SlickLightbox.prototype.destroyPrevious = function() {

      /* Destroys lightboxes currently in DOM. */
      return $('body').children('.slick-lightbox').trigger('destroy.slickLightbox');
    };

    SlickLightbox.prototype.getTransitionDuration = function() {

      /* Detects the transition duration to know when to remove stuff from DOM etc. */
      var duration;
      if (this.transitionDuration) {
        return this.transitionDuration;
      }
      duration = this.modalElement.css('transition-duration');
      if (typeof duration === 'undefined') {
        return this.transitionDuration = 500;
      } else {
        return this.transitionDuration = duration.indexOf('ms') > -1 ? parseFloat(duration) : parseFloat(duration) * 1000;
      }
    };

    return SlickLightbox;

  })();
  defaults = {
    background: 'rgba(0,0,0,.8)',
    closeOnEscape: true,
    closeOnBackdropClick: true,
    destroyTimeout: 500,
    itemSelector: 'a',
    navigateByKeyboard: true,
    src: false,
    caption: false,
    captionPosition: 'dynamic',
    images: false,
    slick: {}
  };
  $.fn.slickLightbox = function(options) {

    /* Fires the plugin. */
    options = $.extend({}, defaults, options);
    new SlickLightbox(this, options);
    return this;
  };
  return $.fn.unslickLightbox = function() {

    /* Removes everything. */
    return $(this).trigger('destroy.slickLightbox');
  };
})(jQuery);
